BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c

```diff
@@ -0,0 +1,1055 @@
+        //
+        // If it's the first time to access variable region in flash, create a guid hob to record
+        // VAR_ADDED type variable info.
+        // Note that as the resource of PEI phase is limited, only store the limited number of
+        // VAR_ADDED type variables to reduce access time.
+        //
+        VarStoreLibData->IndexTable.Length      = 0;
+        VarStoreLibData->IndexTable.StartPtr    = GetStartPointer (VariableStoreHeader);
+        VarStoreLibData->IndexTable.EndPtr      = GetEndPointer   (VariableStoreHeader);
+        VarStoreLibData->IndexTable.GoneThrough = 0;
+        //
+        // Set the Lib data after Ftw and Index table info is updated
+        //
+        Status = SetLibraryData (PcdGet8(PcdUefiVariableLibId), VarStoreLibData, sizeof(UEFI_VAR_STORE_LIBRARY_DATA));
+      } else {
+        StoreInfo->IndexTable = &VarStoreLibData->IndexTable;
+      }
+
+      break;
+
+    default:
+      ASSERT (FALSE);
+      break;
+  }
+
+  StoreInfo->VariableStoreHeader = VariableStoreHeader;
+  return VariableStoreHeader;
+}
+
+/**
+  Get variable header that has consecutive content.
+
+  @param StoreInfo      Pointer to variable store info structure.
+  @param Variable       Pointer to the Variable Header.
+  @param VariableHeader Pointer to Pointer to the Variable Header that has consecutive content.
+
+  @retval TRUE          Variable header is valid.
+  @retval FALSE         Variable header is not valid.
+
+**/
+BOOLEAN
+GetVariableHeader (
+  IN UEFI_VARIABLE_STORE_INFO    *StoreInfo,
+  IN UEFI_VARIABLE_HEADER        *Variable,
+  OUT UEFI_VARIABLE_HEADER       **VariableHeader
+  )
+{
+  EFI_PHYSICAL_ADDRESS            TargetAddress;
+  EFI_PHYSICAL_ADDRESS            SpareAddress;
+  UINTN                           PartialHeaderSize;
+  EFI_STATUS                      LibStatus;
+  UEFI_VAR_STORE_LIBRARY_DATA    *VarStoreLibData;
+
+  if (Variable == NULL) {
+    return FALSE;
+  }
+
+  //
+  // First assume variable header pointed by Variable is consecutive.
+  //
+  *VariableHeader = Variable;
+
+  if (StoreInfo->FtwLastWriteData != NULL) {
+    TargetAddress = StoreInfo->FtwLastWriteData->TargetAddress;
+    SpareAddress = StoreInfo->FtwLastWriteData->SpareAddress;
+    if (((UINTN) Variable > (UINTN) SpareAddress) &&
+        (((UINTN) Variable - (UINTN) SpareAddress + (UINTN) TargetAddress) >= (UINTN) GetEndPointer (StoreInfo->VariableStoreHeader))) {
+      //
+      // Reach the end of variable store.
+      //
+      return FALSE;
+    }
+    if (((UINTN) Variable < (UINTN) TargetAddress) && (((UINTN) Variable + GetVariableHeaderSize (StoreInfo->AuthFlag)) > (UINTN) TargetAddress)) {
+      //
+      // Variable header pointed by Variable is inconsecutive,
+      //
+      LibStatus = GetLibraryData (PcdGet8 (PcdUefiVariableLibId), (VOID **)&VarStoreLibData);
+      if (LibStatus == EFI_NOT_FOUND) {
+        *VariableHeader = &VarStoreLibData->VariableHdr;
```

Comment:
Will we get "EFI_NOT_FOUND"?  The library data should have been set previously.  So it should always return success.  I think the logic here needs adjustment.

Reply:
Used a flag to indicate if VariableHdr is updated in library data. If not, only then take this path.